core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -0,0 +1,635 @@
+                // for followers with no actual change to their metadata, just return group information
+                // for the current generation which will allow them to issue SyncGroup
+                responseCallback(JoinGroupResult(
+                  members = Map.empty,
+                  memberId = memberId,
+                  generationId = group.generationId,
+                  subProtocol = group.protocol,
+                  leaderId = group.leaderId,
+                  errorCode = Errors.NONE.code))
+              }
+            }
+        }
+
+        if (group.is(PreparingRebalance))
+          rebalancePurgatory.checkAndComplete(ConsumerGroupKey(group.groupId))
+      }
+    }
+  }
+
+  def handleSyncGroup(groupId: String,
+                      generation: Int,
+                      memberId: String,
+                      groupAssignment: Map[String, Array[Byte]],
+                      responseCallback: SyncCallback) {
+    if (!isActive.get) {
+      responseCallback(Array.empty, Errors.GROUP_COORDINATOR_NOT_AVAILABLE.code)
+    } else if (!isCoordinatorForGroup(groupId)) {
+      responseCallback(Array.empty, Errors.NOT_COORDINATOR_FOR_GROUP.code)
+    } else {
+      val group = coordinatorMetadata.getGroup(groupId)
+      if (group == null)
+        responseCallback(Array.empty, Errors.ILLEGAL_GENERATION.code)
+      else
+        doSyncGroup(group, generation, memberId, groupAssignment, responseCallback)
+    }
+  }
+
+  private def doSyncGroup(group: GroupMetadata,
+                          generationId: Int,
+                          memberId: String,
+                          groupAssignment: Map[String, Array[Byte]],
+                          responseCallback: SyncCallback) {
+    group synchronized {
+      if (!group.has(memberId)) {
+        responseCallback(Array.empty, Errors.UNKNOWN_MEMBER_ID.code)
+      } else if (generationId != group.generationId) {
+        responseCallback(Array.empty, Errors.ILLEGAL_GENERATION.code)
+      } else {
+        group.currentState match {
+          case Dead =>
+            responseCallback(Array.empty, Errors.UNKNOWN_MEMBER_ID.code)
+
+          case PreparingRebalance =>
+            responseCallback(Array.empty, Errors.REBALANCE_IN_PROGRESS.code)
+
+          case AwaitingSync =>
+            group.get(memberId).awaitingSyncCallback = responseCallback
+            completeAndScheduleNextHeartbeatExpiration(group, group.get(memberId))
+
+            // if this is the leader, then we can transition to stable and
+            // propagate the assignment to any awaiting members
+            if (memberId == group.leaderId) {
+              group.transitionTo(Stable)
+              propagateAssignment(group, groupAssignment)
+            }
+
+          case Stable =>
+            // if the group is stable, we just return the current assignment
+            // TODO: should we return an error if the leader is trying to submit new synchronized state?
```

Comment:
In the current protocol the leader should never submit the sync request twice, or in otherwords submit a sync group request without receiving a join group response; and if they do it would be due to a re-send upon ack failed, etc. Hence I think in the Stable state we should not distinguish leader with others any more and just treat everyone as the same.